firewalldb/kvstores_sql.go

```diff
@@ -0,0 +1,452 @@
+// value is found, nil is returned.
+//
+// NOTE: part of the KVStore interface.
+func (s *sqlKVStore) Get(ctx context.Context, key string) ([]byte, error) {
+	value, err := s.get(ctx, key)
+	if errors.Is(err, sql.ErrNoRows) ||
+		errors.Is(err, session.ErrUnknownGroup) {
+
+		return nil, nil
+	} else if err != nil {
+		return nil, err
+	}
+
+	return value, nil
+}
+
+// Set sets the given key-value pair in the underlying kv store.
+//
+// NOTE: part of the KVStore interface.
+func (s *sqlKVStore) Set(ctx context.Context, key string, value []byte) error {
+	ruleID, sessionID, featureID, err := s.genNamespaceFields(ctx, false)
+	if err != nil {
+		return err
+	}
+
+	// We first need to figure out if we are inserting a new record or
+	// updating an existing one. So first do a GET with the same set of
+	// params.
+	oldValue, err := s.get(ctx, key)
+	if err != nil && !errors.Is(err, sql.ErrNoRows) {
+		return err
+	}
+
+	// No such entry. Add new record.
+	if errors.Is(err, sql.ErrNoRows) {
+		return s.queries.InsertKVStoreRecord(
+			ctx, sqlc.InsertKVStoreRecordParams{
+				EntryKey:  key,
+				Value:     value,
+				Perm:      s.params.perm,
+				RuleID:    ruleID,
+				SessionID: sessionID,
+				FeatureID: featureID,
+			},
+		)
+	}
+
+	// If an entry exists but the value has not changed, there is nothing
+	// left to do.
+	if bytes.Equal(oldValue, value) {
+		return nil
+	}
+
+	// Otherwise, the key exists but the value needs to be updated.
+	switch {
+	case sessionID.Valid && featureID.Valid:
+		return s.queries.UpdateFeatureKVStoreRecord(
+			ctx, sqlc.UpdateFeatureKVStoreRecordParams{
+				Key:       key,
+				Value:     value,
+				Perm:      s.params.perm,
+				SessionID: sessionID,
+				RuleID:    ruleID,
+				FeatureID: featureID,
+			},
+		)
+
+	case sessionID.Valid:
+		return s.queries.UpdateSessionKVStoreRecord(
+			ctx, sqlc.UpdateSessionKVStoreRecordParams{
+				Key:       key,
+				Value:     value,
+				Perm:      s.params.perm,
+				SessionID: sessionID,
+				RuleID:    ruleID,
+			},
+		)
+
+	case featureID.Valid:
+		return fmt.Errorf("a global feature kv store is " +
+			"not currently supported")
+	default:
+		return s.queries.UpdateGlobalKVStoreRecord(
+			ctx, sqlc.UpdateGlobalKVStoreRecordParams{
+				Key:    key,
+				Value:  value,
+				Perm:   s.params.perm,
+				RuleID: ruleID,
+			},
+		)
+	}
+}
+
+// Del deletes the value under the given key in the underlying kv store.
+//
+// NOTE: part of the KVStore interface.
+func (s *sqlKVStore) Del(ctx context.Context, key string) error {
+	// Note: we pass in true here for "read-only" since because this is a
+	// Delete, if the record does not exist, we don't need to create one.
+	// But no need to error out if it doesn't exist.
+	ruleID, sessionID, featureID, err := s.genNamespaceFields(ctx, true)
+	if errors.Is(err, sql.ErrNoRows) ||
+		errors.Is(err, session.ErrUnknownGroup) {
+		fmt.Println(err)
```

Comment:
Nit: Other than the linting issue here, is this an intentional `fmt.Println` and not a `log`?

Reply:
this is the linting issue :) 

anycase, fixed now!